File: 206.反转链表.java
/*
 * @lc app=leetcode.cn id=206 lang=java
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode() {}
 * ListNode(int val) { this.val = val; }
 * ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {

    // /**
    // * 1. 迭代遍历法
    // *
    // * @param head
    // * @return
    // */
    // public ListNode reverseList(ListNode head) {
    // if (head == null)
    // return null;
    // ListNode pre = head;
    // ListNode cur = head.next;
    // ListNode temp;
    // while (cur != null) {
    // temp = cur.next;
    // cur.next = pre;
    // pre = cur;
    // cur = temp;
    // }
    // head.next = null;
    // return pre;
    // }

    /**
     * 递归
     * 
     * @param head
     * @return
     */
    public ListNode reverseList(ListNode head) {
        // 链表只有一个节点时，返回自身
        if (head == null || head.next == null) {
            return head;
        }
        ListNode last = reverseList(head.next);
        // 节点指向反转
        head.next.next = head;
        // 链表结尾指向null
        head.next = null;
        return last;
    }
}
// @lc code=end

File: python/206.反转链表.py
#
# @lc app=leetcode.cn id=206 lang=python
#
# [206] 反转链表
#

# @lc code=start
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution(object):
    def reverseList(self, head):
        
        if not head or not head.next:
            return head
        
        new_head = self.reverseList(head.next)
        
        head.next.next = head
        
        head.next = None
        return new_head
            
        
# @lc code=end

head = ListNode(1)
head.next = ListNode(2)
head.next.next = ListNode(3)
head.next.next.next = ListNode(4)
head.next.next.next.next = ListNode(5)
new_head = Solution().reverseList(head)
while new_head:
    print(new_head.val)
    new_head = new_head.next